### DISCUSSION TITLE
Is there an ability to synchronize which books I have open from device to device (with syncthing or within koreader)?  Same question for synchronization of highlights

### DISCUSSION THREAD

**Discussion by arooni** [koreader/koreader#9189] [Ideas]
I find myself reading on a Kindle Paperwhite, a Samsung phone, a Samsung S8 Ultra tablet, and a Boox Nova Air.

Sometimes I'll start a book on one device, and then totally forget where in my syncthing synchronized directory (same place on all devices) where the book was.

Is there a way I could piggyback on my existing syncthing directory to synchronize some part of the koreader directory (files/folders there) or a feature request I could write that would let me :

1.  Synchronize the history list across each device.  I.e. open a new file on one device, all other devices would receive this information and my history screen on any device would be the same.  
2. Synchronize the annotations, notes and bookmarks from device to device.  i.e. I highlight a passage on one book, then all other devices receive this data.  Would be super handy.  

Many thanks for reading this.

**Comment by poire-z** [comment]
Looks like you have already opened 3 issues requesting this :) #4866 #4780 #4587
(And the answers are no, not yet, not until someone is willing to think about it and implement it.)

**Reply by arooni** [reply]
What about using syncthing and telling it to sync the files themselves that are holding history and highlights?

**Comment by bilogic** [comment]
https://anarc.at/hardware/tablet/kobo-clara-hd/ i managed to sync up using syncthing, it synchronizes the whole books folder eliminating the need to sync multiple different things

**Reply by Frenzie** [reply]
I didn't realize you were facing an x, y problem. Syncthing is by far the simplest solution if that's what you want. ;-)

**Reply by bilogic** [reply]
Yea, I did not expect syncthing to be able to run on the kobo, I found the article after going through a bunch of webdav.

The author mentioned high power usage if syncthing was ran on boot. 

I wonder if syncthing can be made to run just before standby, and killed when kobo is powered up.

Do you have experience to share?

**Reply by Frenzie** [reply]
Other than having tried that out of idle curiosity back in 2016 or 2017 and finding you can set it up in minutes, nope.

If I were interested in sync in this context, I'd use Unison or rsync.

> The author mentioned high power usage.

Just start or kill -STOP/-CONT it manually.

**Reply by bilogic** [reply]
I was editing my last message when your reply came in.

Is there a way to hook into KOReader to run syncthing after power button is pressed to standby, and kill it when kobo is woken?

**Reply by Frenzie** [reply]
For example, a frequently used rsync command might look something like:

```
BRANCH=master; { git diff --name-only; git diff --name-only origin/${BRANCH}; } | sort | uniq | rsync -arz --files-from=- --no-dirs --whole-file --verbose . "root@[ip]/mnt/onboard/.adds/koreader"
```

Though I think @NiLuJe writes it slightly differently.

**Reply by bilogic** [reply]
Thanks for sharing, but why is git required? My experience with rsync is it can be configured to take care of all changes. 

Anyway, all I'm after is just a way to manage books on my devices from the PC (instead of struggling on the device). Syncthing has helped with my phone for many years now.

**Reply by Frenzie** [reply]
> Is there a way to hook into KOReader to run syncthing after power button is pressed to standby, and kill it when kobo is woken?

You can presumably just hook into the Suspend event along these lines, though given that killing wifi is the very first step of suspend because a powered wifi module will prevent it from succeeding I'm pretty sure this won't do much… although, if you combine it with resume as you said then you could simply let it run for a few minutes after resume and kill it then.

```lua
local EventListener = require("ui/widget/eventlistener")
local SyncthingListener = EventListener:extend{}

function SyncthingListener:onSuspend()
    -- Raise syncthing from the dead
end
```

https://github.com/koreader/koreader/pull/9970
https://github.com/koreader/koreader/wiki/User-patches#available_patches

**Reply by bilogic** [reply]
ok thanks, let me try that out

**Reply by Frenzie** [reply]
> Thanks for sharing, but why is git required? My experience with rsync is it can be configured to take care of all changes.

That's exactly why, because otherwise it'd overwrite things I wouldn't want it to overwrite, whether locally on the Kobo or from other branches touching different files I have active simultaneously.

> Anyway, all I'm after is just a way to manage books on my devices from the PC (instead of struggling on the device). Syncthing has helped with my phone for many years now.

I use Thunar, Dolphin, Filezilla, and Directory Opus. `sftp://etc`.

**Reply by bilogic** [reply]
Ok, I guess we have different definitions for `sync` 🤣
Mine is `keep them all the same`.

**Reply by NiLuJe** [reply]
> > Anyway, all I'm after is just a way to manage books on my devices from the PC (instead of struggling on the device).

Err, calibre? ;p

**Reply by bilogic** [reply]
It's too clunky for me, another UI that I need to learn and manage. I just need to manage the files like a thumbdrive for me to drag and drop.

**Comment by notDavid** [comment]
Hi all,

FYI, another possible solution with Syncthing i stumbled upon: https://nicholaslyz.com/blog/2023/10/17/syncing-my-kindle-with-calibre-web-koreader-and-syncthing/

**Comment by eddiermar** [comment]
Hello there... I have been using this plugin for some months now and work as expected in 3 devices (kobo, android, debian x64 server): I found the PDFs highlights, notes, bookmarks and sidecars (SDR folder) synced correctly. The only caveat is that the position of books (last page open) and the full history won't update.

https://github.com/jasonchoimtt/koreader-syncthing

It would be good to find this plugin as official in koreader.

Note: I am not the developer of this plugin, and I am not related to him.

**Reply by Frenzie** [reply]
The current situation seems pretty much perfect. Syncthing isn't something we want to ship because of size, maintenance, and support. A third-party plugin that takes care of all that is fantastic. :-)

But it can be added to https://github.com/koreader/contrib as a submodule so people can find it more easily.

**Reply by notDavid** [reply]
Note that the Plugin does not seem to auto-update Syncthing. Also it has other limitations, i read it only works on local networks.

I followed [this](https://nicholaslyz.com/blog/2023/10/17/syncing-my-kindle-with-calibre-web-koreader-and-syncthing/) guide which uses [this](https://github.com/gutenye/syncthing-kindle) method of installing Syncthing.
This supports both auto-updating and Syncthing works without the limitations.
